lib/model/autocomplete.dart

```diff
@@ -176,17 +164,72 @@ class AutocompleteViewManager {
 ///  * When the object will no longer be used, call [dispose] to free
 ///    resources on the [PerAccountStore].
 class MentionAutocompleteView extends ChangeNotifier {
-  MentionAutocompleteView._({required this.store, required this.narrow});
+  MentionAutocompleteView._({
+    required this.store,
+    required this.narrow,
+    required this.sortedUsers,
+  });
 
   factory MentionAutocompleteView.init({
     required PerAccountStore store,
     required Narrow narrow,
   }) {
-    final view = MentionAutocompleteView._(store: store, narrow: narrow);
+    final users = store.users.values.toList();
+    final sortedUsers = sortByRelevance(users: users, narrow: narrow, store: store);
+    final view = MentionAutocompleteView._(
+      store: store,
+      narrow: narrow,
+      sortedUsers: sortedUsers,
+    );
     store.autocompleteViewManager.registerMentionAutocomplete(view);
     return view;
   }
 
+  static List<User> sortByRelevance({
+    required List<User> users,
+    required Narrow narrow,
+    required PerAccountStore store,
+  }) {
+    switch (narrow) {
+      case StreamNarrow():
+      case TopicNarrow():
+      case DmNarrow():
+        users.sort((userA, userB) => compareByRelevance(
+          userA: userA, userB: userB, store: store));
+      case CombinedFeedNarrow():
+        // do nothing in this case for now
+    }
+    return users;
+  }
+
+  static int compareByRelevance({
+    required User userA,
+    required User userB,
+    required PerAccountStore store,
+  }) {
+    final dmPrecedence = compareByDms(userA, userB, store: store);
+    return dmPrecedence;
+  }
+
+  /// Determines which of the two users is more recent in DM conversations.
+  ///
+  /// Returns a negative number if [userA] is more recent than [userB],
+  /// returns a positive number if [userB] is more recent than [userA],
+  /// and returns `0` if both [userA] and [userB] are equally recent
+  /// or there is no DM exchanged with them whatsoever.
+  static int compareByDms(User userA, User userB, {required PerAccountStore store}) {
```

Comment:
```suggestion
  @visibleForTesting
  static int compareByDms(User userA, User userB, {required PerAccountStore store}) {
```

Conceptually this is a private helper — it's exposed only because that's convenient for unit testing.

(We also don't *have* to expose it in order to write all the tests we like; we could do a more end-to-end test by running a query "" and seeing what order the results come in. But the meaning of this is fairly crisp, so it's basically harmless to expose if it's convenient for tests.)